chore(internal): note schema parity when sqlite backend is enabled

hikchoi opened the pull request:
# chore(internal): note schema parity when sqlite backend is enabled

This PR:
- properly parse and apply note schemas during engine init when `useSqlite` is enabled.
- supersedes #3822

Notes: 
- now `SchemaNotes` table is properly populated after initialization.
- currently schema initialization happens after all delta initialization steps are done. this may change once we move on to a more proper `Schemas` table in the future.
- there are a few points we can optimize (e.g. remove expensive `.get` calls used to create `NoteDicts`), but currently the init step is implemented to use the existing `SchemaUtils.matchDomain` as much as possible with minimal change to the existing way we do it. this could be done by adding sqlite specific utilities to create `NoteDicts` in one query for example. this is also left for future iteration.

# Pull Request Checklist

## Code

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

## Instrumentation

### Basics
- [~] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [~] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## Tests

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [x] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

## Docs
- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

## Close the Loop

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

hikchoi commented on packages/engine-server/src/drivers/sqlite/tables/HierarchyTableUtils.ts:
```diff
@@ -133,4 +133,33 @@ export class HierarchyTableUtils {
+    db: Database,
+    domainId: string
+  ): ResultAsync<HierarchyTableRow[], SqliteError> {
+    const sql = [
```
anyone have a better solution to annotate multiline sql queries?

jonathanyeung replied:
imo this is a pretty good solution. I've started to adopt it in some local changes.

jonathanyeung commented on packages/engine-server/src/drivers/sqlite/tables/HierarchyTableUtils.ts:
```diff
@@ -5,7 +5,7 @@ import { SqliteError } from "../SqliteError";
 import { SqliteQueryUtils } from "../SqliteQueryUtils";
 
 export class HierarchyTableRow {
-  constructor(public parentId: string, public childId: string) {}
```
I had actually renamed from parent/child -> parentId/childId in my first PR in response to a code review comment.  Was it intentional to drop the 'id' part or just a merge artifact?

hikchoi replied:
I changed it because all the queries seem to insert them as `parent` and `child`. I just assumed the tables were defined correctly.

hikchoi replied:
I'll change it back to `parentId` and `childId` and instead change the queries so the column names follow.

jonathanyeung commented on packages/engine-server/src/drivers/sqlite/tables/HierarchyTableUtils.ts:
```diff
@@ -133,4 +133,33 @@ export class HierarchyTableUtils {
+      `  JOIN "cte_hierarchy" c ON c.child = h.parent`, // 5. to the cte until we run out of children
+      `)`,
+      `SELECT * FROM "cte_hierarchy"`, // 6. select everything from the created view.
+    ].join("\n");
```
let's drop the "\n" and just join with a " " separator.

jonathanyeung commented on packages/engine-server/src/drivers/sqlite/tables/HierarchyTableUtils.ts:
```diff
@@ -133,4 +133,33 @@ export class HierarchyTableUtils {
+      `  UNION ALL`, // UNION ALL assuming we don't have dupes in the hierarchy table
+      `  SELECT h.parent, h.child`,
+      `  FROM "Hierarchy" h`, // 4. we are going to join the hierarchy table
+      `  JOIN "cte_hierarchy" c ON c.child = h.parent`, // 5. to the cte until we run out of children
```
Can we add some infinite loop protection here? In case we have a loop between parent-child, child-parent somehow.

hikchoi replied:
Ideally we should do that check before inserting, but yeah. I'll add a check here for now

jonathanyeung commented on packages/engine-server/src/drivers/sqlite/tables/HierarchyTableUtils.ts:
```diff
@@ -133,4 +133,33 @@ export class HierarchyTableUtils {
+   *
+   * more on recursive CTEs: https://www.sqlite.org/lang_with.html
+   */
+  public static getAllInDomain(
```
How about `getAllDescendants()` as a name? Since we're just dealing with a concept of ID's here, I don't think we need to user the word 'domain'.

jonathanyeung commented on packages/engine-server/src/drivers/sqlite/SqliteMetadataStore.ts:
```diff
@@ -30,6 +39,94 @@ import { SchemaNotesTableUtils } from "./tables/SchemaNotesTableUtils";
+   * Goes through all domains and recursively apply schemas.
+   */
+  async initSchema(fileStore: IFileStore, wsRoot: string, logger: DLogger) {
+    // this whole thing to parseAllNoteFilesForSqlite / SqliteDbFactory
```
i think the beginning "// TODO: move" part of the comment got truncated.

hikchoi replied:
right. I'll fix that.

jonathanyeung commented on packages/engine-server/src/drivers/sqlite/SqliteMetadataStore.ts:
```diff
@@ -30,6 +39,94 @@ import { SchemaNotesTableUtils } from "./tables/SchemaNotesTableUtils";
+              })
+            )
+          );
+          // this is too expensive
```
How long does this schema setup take for org-private? (maybe we do this measurement/tuning as a follow up)  If it helps, I left a test-case snippet that can be locally run to get an idea of how long things take here: [[user.jonathan.dev.sqlite-temp-code]]

Might need to add some additional logging though (I also am working on a pr to improve some of our logging in parseAllNoteFiles()

hikchoi replied:
have to test that out. I'll let you know.

hikchoi replied:
Seems like `SqliteMetadataStore.initSchema` take about 2500ms consistenly on my machine with `org-workspace`.

For comparison on my machine it take about 8 seconds to for delta init + schema init to finish. so about 30% of the time is spent on `initSchema`.

hikchoi commented:
update:

- added a limit to the max recursion by limiting the maximum hierarchy depth to 32 (arbitrary). We can extend this to worst case by setting it to the number of rows in the `Hierarchy` table.
- fixed an issue where automatically generated schemas were creating duplicate note -> schema relations in `SchemaNotes` table. I'm doing this by truncating the `SchemaNotes` table each session.

jonathanyeung reviewed (approved):
nice!
